### PR TITLE
Replace MemoryMarshal calls to direct Unsafe calls

### PR DESCRIPTION
**Pull request by vanbukin**
In that places we can call System.Runtime.CompilerServices.Unsafe directly instead of calling MemoryMarshal methods.

### CODE REVIEW

**Comment by jkotas** [comment]
Why is this better? If this is a performance improvement, do you have any numbers for it?

**Comment by EgorBo** [comment]
> Why is this better? If this is a performance improvement, do you have any numbers for it?

I guess it eliminates an extra bounds check
```csharp
var g = new byte[16];
if (BitConverter.IsLittleEndian)
{
    MemoryMarshal.TryWrite<Guid>(g, ref this);
}
```

when `TryWrite` is inlined:

```csharp
var g = new byte[16];
Span<byte> destination = g;
if (16 > (uint)destination.Length)   // this check won't be eliminated 
    return false;                    // destination.Length is always 16
Unsafe.WriteUnaligned<Guid>(ref GetReference(destination), value);
return true;
```
It'd be nice if JIT could do it though (there are issues for this)

**Comment by vanbukin** [comment]
@jkotas benchmark source https://gist.github.com/vanbukin/a5f6f3d294e7b11b8782d79706f9ef75
As @EgorBo say - this PR is just eliminates an extra bounds check by using code from `MemoryMarshal` directly. I think `ToByteArray` is little bit slower cuz I create a new Span. If it necessary maybe I should cast resulting byte array instead of creating a new one.
```BenchmarkDotNet=v0.12.0, OS=macOS 10.15.3 (19D76) [Darwin 19.3.0]
Intel Core i7-8850H CPU 2.60GHz (Coffee Lake), 1 CPU, 12 logical and 6 physical cores
.NET Core SDK=3.1.102
  [Host]     : .NET Core 3.1.2 (CoreCLR 4.700.20.6602, CoreFX 4.700.20.6702), X64 RyuJIT
  Job-SGTUDR : .NET Core 5.0.0 (CoreCLR 42.42.42.42424, CoreFX 42.42.42.42424), X64 RyuJIT

Toolchain=CoreRun
```
`master`
|           Method |     Mean |     Error |    StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|----------------- |---------:|----------:|----------:|-------:|------:|------:|----------:|
|      ToByteArray | 6.195 ns | 0.0762 ns | 0.0712 ns | 0.0085 |     - |     - |      40 B |
|    TryWriteBytes | 5.237 ns | 0.0418 ns | 0.0391 ns |      - |     - |     - |         - |
| CtorReadOnlySpan | 2.642 ns | 0.0240 ns | 0.0213 ns |      - |     - |     - |         - |

`pr`
|           Method |     Mean |     Error |    StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|----------------- |---------:|----------:|----------:|-------:|------:|------:|----------:|
|      ToByteArray | 6.239 ns | 0.0550 ns | 0.0459 ns | 0.0085 |     - |     - |      40 B |
|    TryWriteBytes | 5.102 ns | 0.0392 ns | 0.0348 ns |      - |     - |     - |         - |
| CtorReadOnlySpan | 2.178 ns | 0.0251 ns | 0.0234 ns |      - |     - |     - |         - |

File: src/libraries/System.Private.CoreLib/src/System/Guid.cs [hunk 1]
```diff
@@ -732,7 +732,7 @@ public byte[] ToByteArray()
             var g = new byte[16];
             if (BitConverter.IsLittleEndian)
             {
-                MemoryMarshal.TryWrite<Guid>(g, ref this);
+                Unsafe.WriteUnaligned(ref MemoryMarshal.GetReference(new Span<byte>(g)), this);
```
Comment (EgorBo) [inline src/libraries/System.Private.CoreLib/src/System/Guid.cs]: `Unsafe.WriteUnaligned(ref g[0], this)` ?

**Comment by vanbukin** [comment]
.ToByteArray()

**PR**
```System.Guid.ToByteArray()  
mov     ecx,10h
call    qword ptr [System_Private_CoreLib!<>f__AnonymousType0`1[System.__Canon].get_messag
lea     rdx,[rax+10h]
mov     ecx,dword ptr [rax+8]
movdqu  xmm0,xmmword ptr [rsi]
movdqu  xmmword ptr [rsp+30h],xmm0
movdqu  xmm0,xmmword ptr [rsp+30h]
movdqu  xmmword ptr [rdx],xmm0
add     rsp,40h
```
**master**
```
System.Guid.ToByteArray()  
mov     ecx,10h
call    qword ptr [System_Private_CoreLib!<>f__AnonymousType0`1[System.__Canon].get_messag
lea     rdx,[rax+10h]
mov     ecx,dword ptr [rax+8]
cmp     rcx,10h
jl      System_Private_CoreLib!System.Guid.ToByteArray()$##6000F8F+0x3b
movdqu  xmm0,xmmword ptr [rsi]
movdqu  xmmword ptr [rsp+30h],xmm0
movdqu  xmm0,xmmword ptr [rsp+30h]
movdqu  xmmword ptr [rdx],xmm0
add     rsp,40h
```

**Comment by jkotas** [comment]
Do you see any asm difference for `Guid(ReadOnlySpan<byte> b)` on current master? I believe that the JIT is smart enough to eliminate the bounds check there.

`ToByteArray` is slow method because of it allocates a fresh array to every invocation. Saving two extra instructions by writing more unsafe code is not a good trace-off for it. It is non-measurable improvement. If you would like to fix this one, look into tweaking the JIT to eliminate the bounds check for you.

**Pull request closed without merging**